enforcer/dal/supabase_dal.py

```diff
@@ -0,0 +1,221 @@
+                            "If you store the token in a secret, you must also pass "
+                            "the environment variables ROBUSTA_UI_TOKEN and CLUSTER_NAME to krr-enforcer.\n "
+                        )
+                    try:
+                        decoded = base64.b64decode(token)
+                        return RobustaToken(**json.loads(decoded)), config.global_config.get("cluster_name")
+                    except binascii.Error:
+                        raise Exception(
+                            "binascii.Error encountered. The robusta token provided to Holmes is not a valid base64."
+                        )
+                    except json.JSONDecodeError:
+                        raise Exception(
+                            "json.JSONDecodeError encountered. The Robusta token provided to Holmes could not be parsed as JSON after being base64 decoded."
+                        )
+        return None, None
+
+    def __init_config(self) -> bool:
+        # trying to load the supabase connection parameters from the robusta token, if exists
+        # if not, using env variables as fallback
+        robusta_token, cluster_name = self.__load_robusta_config()
+        if robusta_token:
+            self.account_id = robusta_token.account_id
+            self.url = robusta_token.store_url
+            self.api_key = robusta_token.api_key
+            self.email = robusta_token.email
+            self.password = robusta_token.password
+        else:
+            self.account_id = ROBUSTA_ACCOUNT_ID
+            self.url = STORE_URL
+            self.api_key = STORE_API_KEY
+            self.email = STORE_EMAIL
+            self.password = STORE_PASSWORD
+
+        if cluster_name:
+            self.cluster = cluster_name
+        else:
+            raise Exception("Missing mandatory cluster name")
+
+        # valid only if all store parameters are provided
+        return all([self.account_id, self.url, self.api_key, self.email, self.password])
+
+    def sign_in(self) -> str:
+        logging.info("Supabase DAL login")
+        res = self.client.auth.sign_in_with_password(
+            {"email": self.email, "password": self.password}
+        )
+        self.client.auth.set_session(
+            res.session.access_token, res.session.refresh_token
+        )
+        self.client.postgrest.auth(res.session.access_token)
+        return res.user.id
+
+    def get_latest_krr_scan(self, current_scan_id: Optional[str]) -> (Optional[str], Optional[List[Dict]]):
+        if not self.enabled:
+            return None, None
+
+        try:
+            scans_meta_response = (
+                self.client.table(SCANS_META_TABLE)
+                .select("*")
+                .eq("account_id", self.account_id)
+                .eq("cluster_id", self.cluster)
+                .eq("latest", True)
+                .execute()
+            )
+            if not len(scans_meta_response.data):
```

Comment:
If I understand correctly, this should only return one result. If so, I would verify that.

Reply:
Inserting to this table is atomic and protected: https://github.com/robusta-dev/robusta-storage/blob/235a66aab0719b370aea41759d3eece3b9438069/db/migrations/20240207145819_ScansMeta_state_and_metadata.sql#L10

But I added handling in case there are somehow multiple latest rows